openmc/dagmc.py

```diff
@@ -0,0 +1,593 @@
+    material_overrides : dict
+        A dictionary of material overrides. The keys are material name
+        strings and the values are Iterables of openmc.Material objects. If a 
+        material name is found in the DAGMC file, the material will be replaced
+        with the openmc.Material object in the value.
+    """
+
+    def __init__(self,
+                 filename,
+                 universe_id=None,
+                 name='',
+                 auto_geom_ids=False,
+                 auto_mat_ids=False,
+                 mat_overrides=None):
+        super().__init__(universe_id, name)
+        # Initialize class attributes
+        self.filename = filename
+        self.auto_geom_ids = auto_geom_ids
+        self.auto_mat_ids = auto_mat_ids
+        self.material_overrides = mat_overrides
+        self._dagmc_cells = []
+
+    def __repr__(self):
+        string = super().__repr__()
+        string += '{: <16}=\t{}\n'.format('\tGeom', 'DAGMC')
+        string += '{: <16}=\t{}\n'.format('\tFile', self.filename)
+        return string
+
+    @property
+    def bounding_box(self):
+        with h5py.File(self.filename) as dagmc_file:
+            coords = dagmc_file['tstt']['nodes']['coordinates'][()]
+            lower_left_corner = coords.min(axis=0)
+            upper_right_corner = coords.max(axis=0)
+            return openmc.BoundingBox(lower_left_corner, upper_right_corner)
+
+    @property
+    def filename(self):
+        return self._filename
+
+    @property
+    def material_overrides(self):
+        return self._material_overrides
+
+    @material_overrides.setter
+    def material_overrides(self, val):
+        if val is None:
+            self._material_overrides = val
+            return
+        else:
+            cv.check_type('material overrides', val, dict)
+            for key, value in val.items():
+                # ensuring key is a string and exists in the DAGMC file
+                cv.check_type('material name', key, str)
+                if key not in self.material_names:
+                    raise ValueError(
+                        f"Material name '{key}' not found in DAGMC file")
+                # ensuring overrides is an iterable of material name (strings)
+                cv.check_iterable_type('material objects', value, str)
+
+        self._material_overrides = val
+
+    def add_material_override(self, mat_name, overrides):
+        """Add a material override to the universe.
+
+        Parameters
+        ----------
+        key : str
+            Material name to override
+        value : Iterable of str
+            Material names to replace the key with
+
+        """
+        cv.check_type('material name', mat_name, str)
+        if mat_name not in self.material_names:
+            raise ValueError(
+                f"Material name '{mat_name}' not found in DAGMC file")
+        cv.check_iterable_type('material objects', overrides, str)
+
+        self.material_overrides[mat_name] = overrides
```

Comment:
Does this support a single override as well as a list? Might be convenient to support that if one wants to override a material for all of the cell instances.

Reply:
This should allow to override per cell_id and material name

Reply:
I don't quite understand why it's desirable to have both options. How do we resolve this if both the name of a material and a cell ID that is filled with that material are present in this dictionary?

Reply:
Did this get addressed? Can you explain how the above scenario will be handled?

Reply:
I think I missed understood at first.

I think this is now addressed.